### PR TITLE
fix(ad_service_state_monitor): lock shread variables to prevent race …

### PR DESCRIPTION
**Pull request by ytakano**
`state_machine_` and `state_input_` are shared by multiple callback groups (threads).
So, mutex locks have been applied because it may cause race conditions.

Signed-off-by: Yuuki Takano <yuuki.takano@tier4.jp>

## Description

`AutowareStateMonitorNode::state_machine_` and `AutowareStateMonitorNode::state_input_` are accessed by multiple threads without mutex locks. This may cause race conditions.
So, mutex locks have been applied to the shared variables.

## Pre-review checklist for the PR author

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by taikitanaka3** [comment]
@ytakano 
Thank you for PR I have met similar problem.
Could you write How to reproduce this issue or bug and how to check if that problem is fixed?

**Comment by ytakano** [comment]
I think reproducing these kind of bug is difficult.

This bug should be on of a [Heisenbug](https://en.wikipedia.org/wiki/Heisenbug).
Only thing we can do is reading source code carefully.

**Comment by taikitanaka3** [comment]
@ytakano 
OK , I will test this later.

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (isamu-takagi) [inline system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:233]: Need `lock(lock_state_input_)`?

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (taikitanaka3) [inline system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:186]: @ytakano 
Could you change to use lock and  unlock correctly like[ behavior path planner](https://github.com/autowarefoundation/autoware.universe/blob/b23f4188b26f21ba1f4443be1d1e6819a1747f69/planning/behavior_path_planner/src/behavior_path_planner_node.cpp#L514) inorder not to change previous code much with fix PR?

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (ytakano) [reply system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:233]: Thank you.
You are right.
`lock(lock_state_input_)` is required here.
I will fix it.

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (ytakano) [reply system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:186]: I think `std::lock_guard` is safer than lock/unlock because `std::lock_guard` automatically unlock a mutex.
I recommend to use `std::lock_guard` instead of lock/unlock in the behavior path planner.

**Comment by ytakano** [comment]
I will fix https://github.com/autowarefoundation/autoware.universe/pull/1395#discussion_r926365564 and PR again later.
Thank you for your review!

Thank you @isamu-takagi san.

**Comment by isamu-takagi** [comment]
@ytakano 
Why close PR? It's recommended to change the PR to draft and then update the branch.

**Comment by taikitanaka3** [comment]
@ytakano 
you can continue fixing here to reflect comments

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (isamu-takagi) [reply system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:186]: I think @taikitanaka3  says the lock is only needed under if condition and should be released before following line `response->success = true`.

**Comment by ytakano** [comment]
Sorry for my misunderstanding.
I reopened.

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (isamu-takagi) [reply system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:186]: However, in this case, it returns immediately, so I think it's OK.

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (ytakano) [reply system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:235]: `lock(lock_state_input_);` has been added before `lock(lock_state_machine_)`.

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (taikitanaka3) [inline system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:275]: @ytakano 
you can unlock lock_state_input_ to move `autoware_state = state_machine_->updateState(state_input_);` below

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (taikitanaka3) [inline system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:279]: please name this variable in a bit better way like lock_state_machine

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (taikitanaka3) [reply system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:186]: yes but I want to limit the effective range of lock guard as small as possible

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (taikitanaka3) [reply system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:235]: isn't lock(lock_state_machine_) variable overwrite by lock(lock_state_input_); ?

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (ytakano) [reply system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:186]: Thank you!
I wii fix it.

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (ytakano) [reply system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:279]: I will rename it.

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (ytakano) [reply system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:275]: `lock_state_input_` will be automatically unlocked after `autoware_state = state_machine_->updateState(state_input_);` because of `std::lock_guard`.
`state_machine_->updateState(state_input_)` takes a reference to the `state_input_ `, and the lock is required here.

By the way, `lock_state_input_` can be unlocked after the if statement in the block.
I will unlock it as earlier as possible.

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (ytakano) [reply system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:235]: Thank you.
I will fix it.

File: perception/traffic_light_ssd_fine_detector/lib/include/trt_ssd.hpp
Comment (ytakano) [reply perception/traffic_light_ssd_fine_detector/lib/include/trt_ssd.hpp:18]: All right.
Thank you.

**Comment by ytakano** [comment]
> I suggest this kind of change in your PR to minimize diff
> https://github.com/autowarefoundation/autoware.universe/pull/1399/files

Thank you for your suggestions, but I think it is not reasonable.
Using `std::guard_mutex` or `std::unique_lock` are safer than manual lock/unlock.
Because we are human beeing, we easily forget to unlock locks or free memory spaces.

For example if someone update this code as follows.
```c++
lock();

// someone add if statement
if (condition) {
  // deadlock will occur because of forgetting ulock
  return;
}

unclock();
```

This may cause deadlock.
However, if we use `std::lock_guard`, it can be prevented automatically.

File: perception/traffic_light_ssd_fine_detector/lib/include/trt_ssd.hpp
Comment (ytakano) [reply perception/traffic_light_ssd_fine_detector/lib/include/trt_ssd.hpp:18]: I rebased.
Thank you.

**Comment by taikitanaka3** [comment]
@ytakano 
> Because we are human beeing, we easily forget to unlock locks or free memory spaces.
I agree but this code has two mutex and std::lock_guard  is defined many places with different variable name. That can also causes mistakes too.

File: perception/traffic_light_ssd_fine_detector/lib/include/trt_ssd.hpp
Comment (taikitanaka3) [reply perception/traffic_light_ssd_fine_detector/lib/include/trt_ssd.hpp:18]: This diff is still remaining.

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (taikitanaka3) [inline system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:276]: @ytakano 
I think lock_state_input.unlock(); is not allowed until end of this block.
(    autoware_state = state_machine_->updateState(state_input_);)

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (taikitanaka3) [inline system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:133]: @ytakano 
I think here also needs 
```suggestion
    std::lock_guard<std::mutex> lock(lock_state_input_);
    state_input_.goal_pose = geometry_msgs::msg::Pose::ConstSharedPtr(p);
```

File: system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp
Comment (taikitanaka3) [inline system/ad_service_state_monitor/src/ad_service_state_monitor_node/ad_service_state_monitor_node.cpp:276]:
```suggestion
```

**Comment by ytakano** [comment]
> I agree but this code has two mutex and std::lock_guard is defined many places with different variable name. That can also causes mistakes too. readability is also necessary

Thank you.
I'm not sure `std::lock_guard` worsen readability.
This idea is similar to smart pointers, which automatically deallocate memory spaces, but smart pointers are widely accepted because of memory safety.

As you pointed out, 2 mutexes are a bit complicated, and it may cause deadlock if locks are acquired reversely.

I think it is possible to use only 1 mutex instead of 2.
Should I rewrite it to use only 1 mutex?

**Comment by taikitanaka3** [comment]
@ytakano 
> As you pointed out, 2 mutexes are bit complicated, and it may cause deadlock if locks are acquired reversely.
I think it is possible to use only 1 mutex instead of 2.
Should I rewrite it to use only 1 mutex?

Your re-writing PR makes readability much better I think this PR is ok. Let's wait for CI

**Comment by taikitanaka3** [comment]
@ytakano 
I am going to test this PR launching autoware for 10 min wait for a bit.

**Review by taikitanaka3 — APPROVED** [review]
I tested simulator works 20mins without problem

LGTM
image

**Comment by taikitanaka3** [comment]
@ytakano 
Yout can merge this PR if you think this PR is merge ready

**Comment by taikitanaka3** [comment]
@ytakano 
Can I merge this PR?

**Comment by ytakano** [comment]
@taikitanaka3 
> Can I merge this PR?

Yes! Please.

**Pull request merged**
